### PR TITLE
fix: asynchronously destroy items evicted on clearHistory navigation

### LINKED ISSUE #829: White screen on navigate with clear history parameter
_From @mayureshjadhav on May 26, 2017 3:47_

I am using NativeSctipt 2.5 and nativescript-angular 1.2.0 in my app. 

Scenario is I am navigating from page 1 to page 2 while navigating I am clearing the history but I can see a white screen while moving from page 1 to page 2 . 

What is the issue?

_Copied from original issue: NativeScript/NativeScript#4269_

### PR DESCRIPTION
**Pull request by Buuhuu**
- queue items eligible to be destroyed on clearRefCache instead of destroying them immediately
- finally destroy them on navigatedFrom event

Fixes #829

### CODE REVIEW

**Comment by SvetoslavTsenov** [comment]
run uitests

**Comment by SvetoslavTsenov** [comment]
uitests

**Comment by sis0k0** [comment]
run ci

File: nativescript-angular/router/page-router-outlet.ts [hunk 1]
```diff
@@ -244,6 +250,7 @@ export class PageRouterOutlet { // tslint:disable-line:directive-class-suffix
                 this.locationStrategy._beginBackPageNavigation();
                 this.locationStrategy.back();
             }
+            setTimeout(() => this.destroyQueuedCacheItems());
```
Comment (Buuhuu) [inline nativescript-angular/router/page-router-outlet.ts]: Maybe that shouldn't be in `navigatedFrom` but in `navigatingTo` event handler?

**Comment by sis0k0** [comment]
run ci

**Comment by Buuhuu** [comment]
Any update on that?

**Comment by vakrilov** [comment]
Hi @Buuhuu.
We've been a bit busy with the `3.1` release - so thats the reason for the delay in the review.

**Comment by Ericky14** [comment]
Hello.

I don't wanna bother you guys, but I just wanted to ask, is this going to be merged anytime soon?

This would solve my issue where setting clearHistory to true makes the services not load properly and thus I get errors on the next page if I try to use the services.

**Comment by Buuhuu** [comment]
@Ericky14 feel free to use my repo. Simply use the following reference in your package.json: 

https://github.com/Buuhuu/nativescript-angular.git#781e4026bc5c8addd049cc10ae7d7069bccbc11d

Though this is not forked from master but using a detached HEAD from the 3.0.0 release tag. I have that running in our (almost live) app.

**Comment by sis0k0** [comment]
run ci

File: nativescript-angular/router/page-router-outlet.ts [hunk 2]
```diff
@@ -242,7 +248,10 @@ export class PageRouterOutlet { // tslint:disable-line:directive-class-suffix
         // Add it to the new page
         page.content = componentView;
 
-        page.on("navigatedFrom", (<any>global).Zone.current.wrap((args: NavigatedData) => {
+        page.on(Page.navigatedToEvent, () => setTimeout(() => {
```
Comment (sis0k0) [inline nativescript-angular/router/page-router-outlet.ts]: Heya, do we need the `setTimeout` here?

Comment (Buuhuu) [inline nativescript-angular/router/page-router-outlet.ts]: Technically we don't BUT what I want to achieve is to push the actually execution to the end of the event loop so when the view tree to destroy is huge the main thread has time to actually do other things before the clean up happens.

**Comment by sis0k0** [comment]
run ci

**Comment by sis0k0** [comment]
LGTM, will merge after green build :smile:

**Comment by Buuhuu** [comment]
We have that (almost) in production already ;)

**Comment by sis0k0** [comment]
sdkwebpack

**Comment by sis0k0** [comment]
sdkwebpack

**Comment by vchimev** [comment]
sdkwebpack

**Comment by berchik** [comment]
Still experiencing the same problem with v4.2. 
When using clearHistory with navigate transition there is a white blank background.
Tested with 4.2 / 4.3.0-rc- / 4.3.0-2017-8-17-1

File: nativescript-angular/router/page-router-outlet.ts (hunk 2 above)
Comment (jogboms) [reply nativescript-angular/router/page-router-outlet.ts]: Good point.

**Pull request merged**